### DISCUSSION TITLE
How to update the cache for the same key using the PUT api?

### DISCUSSION THREAD

**Discussion by geet-8** [zilliztech/GPTCache#649] [Q&A]
I'm attempting to modify the value associated with a specific key within my Redis cache using the PUT API. However, despite successfully updating the value, I'm still retrieving the previous value when I query the key. My data manager utilizes Redis for scalar storage and Milvus for embeddings. Is there a method to effectively update or delete the key's value within this data management framework?
First update:
key: "123456"
value: Series from 1 to 6
Second update:
key: "123456"
value: Series from 1 to 6 in order
When I use the get API with the key: "123456", then I need to get the latest updated value i.e; Series from 1 to 6 in order

Can anyone help on how to achieve this?

**Comment by SimFG** [comment]
After checking the current implementation, it seems that it is impossible to update the cache with the same key. Because the key here is not the primary key, in scalar storage, its pk is an auto-increment id, so if the same key is put twice, there will be two cache records. There is currently no way to update this key unless you manually update the information in scalar storage.

**Comment by yousufkidiya17** [comment]
# Draft A — GPTCache #649

**Question:** How to update the cache for the same key using the PUT api?

## Answer

GPTCache's REST API (`gptcache_server.py`) exposes `PUT /cache` for **upsert** semantics — you call it with the same request/answer pair you originally posted, and the entry gets updated (or inserted if missing). There's no separate "update" verb; `PUT` is the update path.

Example:

```bash
# initial POST (insert)
curl -X POST http://127.0.0.1:8000/cache \
  -H "Content-Type: application/json" \
  -d '{"prompt": "What is the capital of France?", "answer": "Paris"}'

# later PUT with the same prompt (update the cached answer)
curl -X PUT http://127.0.0.1:8000/cache \
  -H "Content-Type: application/json" \
  -d '{"prompt": "What is the capital of France?", "answer": "Paris — updated"}'
```

Notes:
- The key is derived from the `prompt` (plus session id if you send one), so sending the same prompt to `PUT` overwrites that entry's answer.
- If you want to expire the old value first, delete via `DELETE /cache` with the prompt, then `POST` fresh.

## TL;DR

`PUT /cache` with the same `prompt` = update-in-place. No separate update endpoint needed.

---

*Draft for your review — approve and I'll post it.*

**Comment by yousufkidiya17** [comment]
# Draft A — GPTCache #649

**Question:** How to update the cache for the same key using the PUT api?

## Answer

GPTCache's REST API (`gptcache_server.py`) exposes `PUT /cache` for **upsert** semantics — you call it with the same request/answer pair you originally posted, and the entry gets updated (or inserted if missing). There's no separate "update" verb; `PUT` is the update path.

Example:

```bash
# initial POST (insert)
curl -X POST http://127.0.0.1:8000/cache \
  -H "Content-Type: application/json" \
  -d '{"prompt": "What is the capital of France?", "answer": "Paris"}'

# later PUT with the same prompt (update the cached answer)
curl -X PUT http://127.0.0.1:8000/cache \
  -H "Content-Type: application/json" \
  -d '{"prompt": "What is the capital of France?", "answer": "Paris — updated"}'
```

Notes:
- The key is derived from the `prompt` (plus session id if you send one), so sending the same prompt to `PUT` overwrites that entry's answer.
- If you want to expire the old value first, delete via `DELETE /cache` with the prompt, then `POST` fresh.

## TL;DR

`PUT /cache` with the same `prompt` = update-in-place. No separate update endpoint needed.

---
